azure-kusto-data/tests/setup.rs

```diff
@@ -0,0 +1,91 @@
+#![cfg(feature = "mock_transport_framework")]
+use azure_core::auth::{TokenCredential, TokenResponse};
+use azure_core::Error as CoreError;
+use azure_kusto_data::prelude::*;
+use chrono::Utc;
+use dotenv::dotenv;
+use oauth2::AccessToken;
+use std::error::Error;
+use std::path::Path;
+use std::sync::Arc;
+
+pub struct DummyCredential {}
+
+#[async_trait::async_trait]
+impl TokenCredential for DummyCredential {
+    async fn get_token(&self, _resource: &str) -> Result<TokenResponse, CoreError> {
+        Ok(TokenResponse::new(
+            AccessToken::new("some dummy token".to_string()),
+            Utc::now(),
+        ))
+    }
+}
+
+pub async fn create_kusto_client(
+    transaction_name: &str,
+) -> Result<(KustoClient, String), Box<dyn Error + Send + Sync>> {
+    let transaction_path = Path::new(&workspace_root().unwrap())
+        .join(format!("test/transactions/{}", transaction_name));
+    std::fs::create_dir_all(&transaction_path).unwrap();
+    let db_path = transaction_path.join("_db");
+
+    let (service_url, credential, database): (String, Arc<dyn TokenCredential>, String) =
+        if std::env::var(azure_core::mock::TESTING_MODE_KEY).as_deref()
+            == Ok(azure_core::mock::TESTING_MODE_RECORD)
+        {
+            dotenv().ok();
+
+            let client_id =
+                std::env::var("AZURE_CLIENT_ID").expect("Set env variable AZURE_CLIENT_ID first!");
+            let client_secret = std::env::var("AZURE_CLIENT_SECRET")
+                .expect("Set env variable AZURE_CLIENT_SECRET first!");
+            let tenant_id =
+                std::env::var("AZURE_TENANT_ID").expect("Set env variable AZURE_TENANT_ID first!");
+            let service_url = std::env::var("KUSTO_SERVICE_URL")
+                .expect("Set env variable KUSTO_SERVICE_URL first!");
+            let database =
+                std::env::var("KUSTO_DATABASE").expect("Set env variable KUSTO_DATABASE first!");
+
+            // Wee need to persist the database name as well, since it may change per recording run depending on who
+            // records it, is part of the request, and as such validated against.
+            std::fs::write(db_path, &database).unwrap();
+
+            let credential = Arc::new(ClientSecretCredential::new(
+                tenant_id.to_string(),
+                client_id.to_string(),
+                client_secret.to_string(),
+                TokenCredentialOptions::default(),
+            ));
+            (service_url, credential, database)
+        } else {
+            let credential = Arc::new(DummyCredential {});
+            let database = String::from_utf8_lossy(&std::fs::read(db_path).unwrap()).to_string();
+            (String::new(), credential, database)
+        };
+
+    let options = KustoClientOptions::new_with_transaction_name(transaction_name.to_string());
```

Comment:
Can you explain a bit how the transactions work?
Do we actually send and receive data from the server, or is it all mocked with the request/response json files?
If it's all mocked, why do we need an actual client? Does it really count as e2e?

Reply:
The framework works in two modes - the RECORD and REPLAY mode. In Record mode there is just another pipeline element, that does the recording. This has to be executed locally against an actual service. Tests are then run in REPLAY mode, where no more communication with the server happens. However all steps are executed up until the transport level, where we do the communication with the service. Personally I would at least consider that e2e-ish :).

Reply:
But ... does the Kusto team operate a cluster for use in CI/CD? if so that would be real end to end :)

Reply:
We do :)
I'll implement it at some point